src/refiners/training_utils/trainer.py

```diff
@@ -345,11 +349,24 @@ def backward(self) -> None:
         """Backward pass on the loss."""
         self._call_callbacks(event_name="on_backward_begin")
         scaled_loss = self.loss / self.clock.num_step_per_iteration
-        backward(tensors=scaled_loss)
+        if self.scaler is not None:
+            self.scaler.scale(scaled_loss).backward()  # type: ignore
+        else:
+            backward(tensors=scaled_loss)
         self._call_callbacks(event_name="on_backward_end")
         if self.clock.is_optimizer_step:
             self._call_callbacks(event_name="on_optimizer_step_begin")
-            self.optimizer.step()
+            if self.scaler is not None:
```

Comment:
My rule is that you should try not to fold code more than 3 times.

In this case, you can write a separate method `optimizer_step`

```py
def optimizer_step(self) -> None:
    if self.scaler is None:
        self.optimizer.step()
        return
    # rest of the logic here
```